experimental: add valgrind test to flux-sched, fix some leaks

grondo opened the pull request:
This somewhat experimental PR adds the `t5000-valgrind.t` to flux-sched, and fixes some of the leaks found. The valgrind test has the following differences from flux-core:
 * runs with `test_expect_failure` because there are still many leaks. Once all detected leaks are fixed the test can be updated.
 * A hacky attempt to determine if flux-broker was built with the `dlclose` valgrind wrapper is added, and the test is skipped if the symbol `ZZ_Za_dlclose` isn't found.
 * The path to `flux-broker` is determined via `$(PATH=${FLUX_EXEC_PATH} /usr/bin/which flux-broker)`. This should be the flux-broker for the version of flux-core under which flux-sched was configured.
 * An extra test is added to ensure the found path to `flux-broker` is executable.

The `valgrind-workload.sh` is also just a copy from flux-core, this could be extended for flux-sched.

The leaks fixed here were all fairly obvious given valgrind output in #327 . One thing that happened was when shortjson.h was switched to shortjansson.h, the fact that `Jtostr()` now required its argument to be freed was missed. I would actually suggest that the functions in shortjansson.h should be avoided on future development in flux-sched.

This isn't ready for merge yet. I may need to iterate on the valgrind test implementation.

grondo commented:
@dongahn: There's one last minor flux-sched leak detected by valgrind for which I think I'll need your help:
```
==2426630== 40 (24 direct, 16 indirect) bytes in 1 blocks are definitely lost in loss record 23 of 45
==2426630==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==2426630==    by 0x8C6D72D: xzmalloc (xzmalloc.c:39)
==2426630==    by 0x8C66371: partition_new (rs2rank.c:114)
==2426630==    by 0x8C66371: rs2rank_tab_update (rs2rank.c:225)
==2426630==    by 0x8C66EBB: rsreader_hwloc_load (rsreader.c:276)
==2426630==    by 0x8C62BCA: build_hwloc_rs2rank (sched.c:671)
==2426630==    by 0x8C65AB6: load_resources (sched.c:773)
==2426630==    by 0x8C65AB6: mod_main (sched.c:2296)
==2426630==    by 0x40AB18: module_thread (module.c:158)
==2426630==    by 0x59D1183: start_thread (pthread_create.c:312)
==2426630==    by 0x5EE537C: clone (clone.S:111)
==2426630== 
```

Otherwise, once a couple libjsc leak fixes are landed in flux-core, this valgrind test at least will run clean for flux-sched.

dongahn commented:
@grondo: My patch for partition leak is at https://github.com/flux-framework/flux-sched/commit/376a694968993f700f14fe955921c4ddd537bf6b

I wasn't sure how to test this with your `t5000-valgrind.t` so I will appreciate if you can cherry pick and test this under your environment.

grondo commented:
Thanks @dongahn, your fix appeared to work, and along with jsc fixes in core, the valgrind test runs clean:
```
PASS: t5000-valgrind.t 1 - found executable flux-broker
XPASS: t5000-valgrind.t 2 - valgrind reports no new errors on single broker run # TODO known breakage vanished
```

I'll update this PR with your fix, plus I will remove the `test_expect_failure` from the test so that future leaks which are detected and cause testsuite to fail.

grondo commented:
Ok, added @dongahn's fix and updated valgrind test to expect success.

 FYI this will fail travis (or should fail), until flux-framework/flux-core#1494 is merged.

dongahn commented:
LGTM across the board. One question. shortjansson.h is only used in sched? Not in core?

dongahn commented:
flux-framework/flux-core#1494 has been merged. So, restarted the builds. When Travis reports, I will merge.

grondo commented:
>  One question. shortjansson.h is only used in sched? Not in core?

Yes, I take the blame for that one. This replacement of `shortjson.h` with `shortjansson.h` was a bad shortcut taken in e282c891130578672e878e7ad9f94638184929c1

grondo commented:
>  We'll ultimately convert the whole thing to jansson in a near future.

Thanks, good plan. For new code shortjson/shortjansson.h should be avoided.

grondo commented:
Thanks!

dongahn commented:
Thank YOU!

dongahn commented:
@grondo: I merged your change into my dev branch and somehow `t5000-valgrind.t` is failing on quartz. Is there a way to see the valgrind output from this test? 

```
ERROR: t5000-valgrind.t - missing test plan
ERROR: t5000-valgrind.t - exited with status 1
```

grondo commented:
Sorry @dongahn, I missed this comment until now. Try:

```
$ cd t
$ make check TESTS=t5000-valgrind.t FLUX_TESTS_LOGFILE=t debug=t
```

Then look in `t5000-valgrind.output`

grondo commented:
Sorry running the tests in `flux-sched` and saving the output is a bit convoluted, but this is because there is *a lot* of environment setup in `t/Makefile.am` for the tests. We should consider how to make it so that tests can by run by hand as in flux-core, e.g.
```
./t5000-valgrind.t -d -v
```
